### ISSUE TITLE
Remove caching from Prelude while caches are α-normalized

### ISSUE THREAD

**Issue by sjakobi** [dhall-lang/dhall-lang#733]
The Prelude currently uses hashes to cache local imports, for example [here](https://github.com/dhall-lang/dhall-lang/blob/98a341107be8bf5c3814c848e158c7813f69be8f/Prelude/Bool/package.dhall).

The downside of these hashes is that the protected imports are α-normalized, making them [harder to read](https://github.com/dhall-lang/dhall-haskell/issues/1212).

I suggest that until we have standardized a caching system that doesn't cause α-normalization, we remove these hashes.

**Comment by Gabriel439** [comment]
So I don't think we should remove the internal hashes.  They currently provide a lot of value by improving the performance of unprotected imports of the Prelude

However, I am open to the idea of just not α-normalizing imports when caching them or computing the integrity check.  It would be a breaking change, but I think we can figure out a way to ease the migration (like the Haskell implementation providing backwards-compatible support for the old hashes).

**Comment by singpolyma** [comment]
I would be *strongly* opposed to reducing the amount of alpha-normalization we do.  In `dhall-ruby` I actually alpha-normalize every time I beta-normalize already.  Removing this for computing the integrity check would be a big step backwards, in my opinion, since just changing a variable name would change the hash, thus significantly breaking the benefit of the "semantic" in semantic integrity check.

**Comment by Gabriel439** [comment]
@singpolyma: I think my comment in https://github.com/dhall-lang/dhall-haskell/issues/1212#issuecomment-529435584 addresses half of your concern, by proposing a separate type of hash (i.e. `cache:…`) that behaves differently from the current semantic integrity check.

The idea is that if you tag an import with `cache:…` then it does not α-normalize the expression but it also does not reject a hash mismatch.  In other words, the hash is used entirely for caching purposes and not security reasons.

**Comment by singpolyma** [comment]
Would this imply that regular hashes are not cached?  Don't we need to 
protect cached things against cache poisoning?

Since the concern is readability in the output of a single use case, could 
the naming info not be tracked for that use case seperately from the spec?

**Comment by Gabriel439** [comment]
@singpolyma: Regular hashes would still be cached, too.  Both types of hashes would still be protected against cache poisoning because they would still be content-addressable.  Specifically, the `cache:…` would both hash and cache the non-α-normalized form to avoid cache poisoning.

However, the more general reasoning behind the `cache:…` idea is that we currently don't capture the user's intent when we add integrity checks: sometimes they add them for security reasons and sometimes they add them for caching reasons, and by preserving that high-level intent in the code we can design around it (both in the language and in supporting tooling).

The main reason I haven't thought through a Haskell-implementation-specific fix is that it leads to an observable deviation from the standard, which users might interpret as bugs in other implementations.  One of the things I'm actively avoiding is the Haskell implementation becoming the *de facto* Dhall standard in the same way that GHC became the  *de facto* standard for the Haskell.

**Comment by singpolyma** [comment]
>The main reason I haven't thought through a Haskell-implementation-specific 
>fix is that it leads to an observable deviation from the standard, which 
>users might interpret as bugs in other implementations.

Currently, the standard already allows returning non-alpha-normalized 
expressions from an integrity protected import (this is often, but not 
always, done when the import is not cached) so I don't think an 
implementation choosing to preserve names would be a deviation from the 
standard, though it may not be the most obvious first implementation.

**Comment by sjakobi** [comment]
> I don't think an implementation choosing to preserve names would be a deviation from the standard

I don't see how the standard could be interpreted this way. It says

> Cache the fully resolved, αβ-normalized expression, and encoded expression

… and later on

> If the file exists and is readable, verify the file's byte contents match the hash and then decode the expression from the bytes using the decode judgment instead of importing the expression

In fact, I think it would be desirable to have a "transparent" cache, where a protected import is resolved to the exactly same expression, no matter whether it is in the cache or not.

**Comment by singpolyma** [comment]
@sjakobi indeed, the cache should be as transparent as possible (which is one reason why in my implementation I always alpha-normalize even when there is no cache) however this is just as easily an argument in *favour* of name-preservation, since implementations are already allowed to return non-alpha-normalized from a protected expression that is not cached, also doing that when it is cached would not break the standard in any way.
